File: SSMConnectKit/Sources/SSMConnectDomain/PortableProfile.swift
import Foundation

/// The portable connection-profile document (spec §6.1, `contracts/connection-profile.schema.json`).
///
/// This is deliberately a **separate type from `ConnectionProfile`**, not a `Codable` conformance on
/// it. `ConnectionProfile`'s synthesized coding is the app's private `UserDefaults` representation
/// and is free to change; this document is a cross-client contract and is not. Two concrete places
/// where they already disagree, either of which would have shipped a schema-violating export:
///
/// - `ConnectAction` has no raw value, so synthesized encoding emits `{"dcvViewer":{}}` where the
///   schema requires the string `"dcvViewer"`.
/// - The document carries `schemaVersion`, which the native model has no reason to hold.
public struct PortableProfileDocument: Equatable, Sendable {
    /// The only version this client writes, and the only one it accepts (schema: `const: 1`).
    public static let currentSchemaVersion = 1

    /// The one `connectAction` v1 defines.
    public static let dcvViewerAction = "dcvViewer"

    public var schemaVersion: Int = currentSchemaVersion
    public var id: UUID
    public var name: String
    public var ssoStartUrl: String
    public var ssoRegion: String
    public var accountId: String
    public var roleName: String
    public var resourceRegion: String
    public var instanceTagKey: String
    public var instanceTagValue: String
    public var connectMode: ConnectMode?
    public var connectAction: String = dcvViewerAction
    public var localPort: Int
    public var remotePort: Int
    public var agentRemotePort: Int?
    public var secretId: String?

    /// Optional properties this client does not model, preserved verbatim so a document written by
    /// a newer or different client survives a round-trip through this one ("unknown optional
    /// properties are permitted and SHOULD be preserved on round-trip").
    ///
    /// Preservation is not unconditional: `ProfilePortability` rejects a document carrying any key
    /// on the schema's credential deny-list before it reaches here, so this can never become a
    /// smuggling channel for secrets.
    public var unknownFields: [String: JSONValue] = [:]
}

// MARK: - Mapping to and from the app's native model

public extension PortableProfileDocument {
    /// Build the portable document for a stored profile.
    ///
    /// - Parameter unknownFields: optional properties carried in from a previous import, so
    ///   export(import(x)) preserves them.
    public init(profile: ConnectionProfile, unknownFields: [String: JSONValue] = [:]) {
        self.id = profile.id
        self.name = profile.name
        self.ssoStartUrl = profile.ssoStartUrl
        self.ssoRegion = profile.ssoRegion
        self.accountId = profile.accountId
        self.roleName = profile.roleName
        self.resourceRegion = profile.resourceRegion
        self.instanceTagKey = profile.instanceTagKey
        self.instanceTagValue = profile.instanceTagValue
        self.connectMode = profile.connectMode
        self.localPort = profile.localPort
        self.remotePort = profile.remotePort
        self.agentRemotePort = profile.agentRemotePort
        self.secretId = profile.secretId
        self.unknownFields = unknownFields
    }

    /// The stored profile this document describes.
    ///
    /// `connectMode` and `agentRemotePort` stay optional rather than being resolved to their
    /// defaults here: the schema says absent means `singleUser` / 8444, and `ConnectionProfile`
    /// already encodes exactly that through `resolvedConnectMode` / `resolvedAgentRemotePort`.
    /// Resolving eagerly would rewrite a legacy profile's absent field into a present one and
    /// change what a re-export looks like.
    public var profile: ConnectionProfile {
        ConnectionProfile(
            id: id,
            name: name,
            ssoStartUrl: ssoStartUrl,
            ssoRegion: ssoRegion,
            accountId: accountId,
            roleName: roleName,
            resourceRegion: resourceRegion,
            instanceTagKey: instanceTagKey,
            instanceTagValue: instanceTagValue,
            secretId: secretId,
            localPort: localPort,
            remotePort: remotePort,
            connectAction: .dcvViewer,
            connectMode: connectMode,
            agentRemotePort: agentRemotePort
        )
    }
}

// MARK: - Errors

/// Why a portable profile document was rejected.
///
/// Every case names the offending field and value, for the same reason `ProfileConfigError` does:
/// an import failure the user cannot act on is barely better than a silent one.
public enum ProfilePortabilityError: LocalizedError, Equatable {
    case malformedJSON
    case notAnObject
    case unsupportedSchemaVersion(found: Int)
    case missingField(String)
    case invalidField(field: String, value: String, reason: String)
    case forbiddenField(String)

    public var errorDescription: String? {
        switch self {
        case .malformedJSON:
            "This file isn't valid JSON, so it can't be read as a connection profile."
        case .notAnObject:
            "A connection profile must be a single JSON object."
        case let .unsupportedSchemaVersion(found):
            "This profile uses format version \(found), but this version of SSM Connect only "
                + "understands version \(PortableProfileDocument.currentSchemaVersion). Update the app, "
                + "or export the profile again from a matching version."
        case let .missingField(field):
            "The profile is missing the required field \"\(field)\"."
        case let .invalidField(field, value, reason):
            "The profile's \"\(field)\" value \(value.isEmpty ? "\"\"" : "\"\(value)\"") is not valid: \(reason)"
        case let .forbiddenField(field):
            "The profile carries a \"\(field)\" field. A portable profile carries identifiers only, "
                + "never credentials or secrets — this file was not produced by SSM Connect and was not imported."
        }
    }
}

// MARK: - Encoding and decoding

/// Reads and writes the portable profile document (spec §6.1, AC-03).
///
/// The validation here intentionally duplicates `connection-profile.schema.json` rather than
/// deferring to it: the schema is checked in CI by `contracts/validate.py`, but a shipped client
/// has no JSON Schema validator and still must refuse a bad document with an actionable message.
/// `PortableProfileTests` pins the two together by round-tripping the real contract fixtures.
public enum ProfilePortability {
    /// Keys the schema's security clause forbids outright: live authentication material and
    /// machine-local paths (§6.1, §12). Rejected even though they would otherwise be "unknown
    /// optional properties", because preserving them is exactly the failure mode to avoid.
    public static let forbiddenKeys: Set<String> = [
        "accessKeyId", "secretAccessKey", "sessionToken", "credentials",
        "accessToken", "refreshToken", "ssoAccessToken",
        "password", "dcvPassword", "secretValue", "authToken", "presignedUrl",
        "sessionResponse", "ssmSession", "connectionFilePath", "tempFilePath",
    ]

    /// Properties this client models. Anything else is an unknown optional property.
    private static let knownKeys: Set<String> = [
        "schemaVersion", "id", "name", "ssoStartUrl", "ssoRegion", "accountId", "roleName",
        "resourceRegion", "instanceTagKey", "instanceTagValue", "connectMode", "connectAction",
        "localPort", "remotePort", "agentRemotePort", "secretId",
    ]

    // MARK: Export

    /// Encode a profile as a portable document.
    ///
    /// Keys are sorted and the output is pretty-printed: an exported profile is a file a human
    /// diffs and a reviewer reads, and stable key order is what makes that useful.
    public static func export(_ profile: ConnectionProfile, unknownFields: [String: JSONValue] = [:]) throws -> Data {
        let document = PortableProfileDocument(profile: profile, unknownFields: unknownFields)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(document.jsonObject())
    }

    /// Encode a profile as a portable document string.
    public static func exportString(_ profile: ConnectionProfile, unknownFields: [String: JSONValue] = [:]) throws -> String {
        String(decoding: try export(profile, unknownFields: unknownFields), as: UTF8.self)
    }

    // MARK: Import

    /// Decode and validate a portable document.
    ///
    /// Validation is total: the returned document has already been checked against every constraint
    /// this client can express, so callers do not re-validate.
    public static func importDocument(from data: Data) throws -> PortableProfileDocument {
        let value: JSONValue
        do {
            value = try JSONDecoder().decode(JSONValue.self, from: data)
        } catch {
            throw ProfilePortabilityError.malformedJSON
        }
        guard case let .object(object) = value else {
            throw ProfilePortabilityError.notAnObject
        }

        // Version first: a document from a future schema must fail on the version, not on whichever
        // field happens to have changed shape.
        let version = try requiredInt(object, "schemaVersion")
        guard version == PortableProfileDocument.currentSchemaVersion else {
            throw ProfilePortabilityError.unsupportedSchemaVersion(found: version)
        }

        // Security boundary before anything else is trusted.
        for key in object.keys where forbiddenKeys.contains(key) {
            throw ProfilePortabilityError.forbiddenField(key)
        }

        let idText = try requiredString(object, "id")
        guard let id = UUID(uuidString: idText) else {
            throw ProfilePortabilityError.invalidField(field: "id", value: idText, reason: "it is not a UUID.")
        }

        let connectMode = try optionalConnectMode(object)
        let secretId = try optionalNullableString(object, "secretId")
        if connectMode == .multiUser, let secretId {
            throw ProfilePortabilityError.invalidField(
                field: "secretId",
                value: secretId,
                reason: "a multi-user workstation authenticates by identity and never uses a shared "
                    + "password, so it must not carry a secret ID."
            )
        }

        let action = try optionalString(object, "connectAction") ?? PortableProfileDocument.dcvViewerAction
        guard action == PortableProfileDocument.dcvViewerAction else {
            throw ProfilePortabilityError.invalidField(
                field: "connectAction",
                value: action,
                reason: "this version only supports \"\(PortableProfileDocument.dcvViewerAction)\"."
            )
        }

        var document = PortableProfileDocument(
            schemaVersion: version,
            id: id,
            name: try requiredNonEmptyString(object, "name"),
            ssoStartUrl: try requiredNonEmptyString(object, "ssoStartUrl"),
            ssoRegion: try requiredRegion(object, "ssoRegion"),
            accountId: try requiredAccountId(object),
            roleName: try requiredNonEmptyString(object, "roleName"),
            resourceRegion: try requiredRegion(object, "resourceRegion"),
            instanceTagKey: try requiredNonEmptyString(object, "instanceTagKey"),
            // Deliberately allowed to be empty: an exported but not-yet-configured profile is valid.
            instanceTagValue: try requiredString(object, "instanceTagValue"),
            connectMode: connectMode,
            connectAction: action,
            localPort: try requiredPort(object, "localPort"),
            remotePort: try requiredPort(object, "remotePort"),
            agentRemotePort: try optionalPort(object, "agentRemotePort"),
            secretId: secretId
        )

        document.unknownFields = object.filter { !knownKeys.contains($0.key) }

        return document
    }

    /// Decode and validate a portable document, returning the profile it describes.
    public static func importProfile(from data: Data) throws -> ConnectionProfile {
        try importDocument(from: data).profile
    }

    // MARK: Field readers

    private static func requiredValue(_ object: [String: JSONValue], _ key: String) throws -> JSONValue {
        guard let value = object[key], value != .null else {
            throw ProfilePortabilityError.missingField(key)
        }
        return value
    }

    private static func requiredString(_ object: [String: JSONValue], _ key: String) throws -> String {
        guard case let .string(text) = try requiredValue(object, key) else {
            throw ProfilePortabilityError.invalidField(field: key, value: "", reason: "it must be a string.")
        }
        return text
    }

    private static func requiredNonEmptyString(_ object: [String: JSONValue], _ key: String) throws -> String {
        let text = try requiredString(object, key)
        guard !text.isEmpty else {
            throw ProfilePortabilityError.invalidField(field: key, value: text, reason: "it must not be empty.")
        }
        return text
    }

    private static func optionalString(_ object: [String: JSONValue], _ key: String) throws -> String? {
        guard let value = object[key], value != .null else { return nil }
        guard case let .string(text) = value else {
            throw ProfilePortabilityError.invalidField(field: key, value: "", reason: "it must be a string.")
        }
        return text
    }

    /// `secretId` is explicitly nullable, so an explicit `null` is valid and means "absent".
    private static func optionalNullableString(_ object: [String: JSONValue], _ key: String) throws -> String? {
        try optionalString(object, key)
    }

    private static func requiredInt(_ object: [String: JSONValue], _ key: String) throws -> Int {
        guard case let .number(number) = try requiredValue(object, key), number == number.rounded() else {
            throw ProfilePortabilityError.invalidField(field: key, value: "", reason: "it must be a whole number.")
        }
        return Int(number)
    }

    private static func requiredRegion(_ object: [String: JSONValue], _ key: String) throws -> String {
        let value = try requiredString(object, key)
        guard AWSRegion.isValid(value) else {
            throw ProfilePortabilityError.invalidField(
                field: key,
                value: value,
                reason: "it is not a syntactically valid AWS region such as eu-central-1."
            )
        }
        return value
    }

    private static func requiredAccountId(_ object: [String: JSONValue]) throws -> String {
        let value = try requiredString(object, "accountId")
        guard value.count == 12, value.allSatisfy(\.isASCII), value.allSatisfy(\.isNumber) else {
            throw ProfilePortabilityError.invalidField(
                field: "accountId",
                value: value,
                reason: "an AWS account ID is exactly 12 digits."
            )
        }
        return value
    }

    private static func requiredPort(_ object: [String: JSONValue], _ key: String) throws -> Int {
        let value = try requiredInt(object, key)
        guard (1...65535).contains(value) else {
            throw ProfilePortabilityError.invalidField(
                field: key,
                value: String(value),
                reason: "a port must be between 1 and 65535."
            )
        }
        return value
    }

    private static func optionalPort(_ object: [String: JSONValue], _ key: String) throws -> Int? {
        guard let value = object[key], value != .null else { return nil }
        guard case let .number(number) = value, number == number.rounded() else {
            throw ProfilePortabilityError.invalidField(field: key, value: "", reason: "it must be a whole number.")
        }
        let port = Int(number)
        guard (1...65535).contains(port) else {
            throw ProfilePortabilityError.invalidField(
                field: key,
                value: String(port),
                reason: "a port must be between 1 and 65535."
            )
        }
        return port
    }

    private static func optionalConnectMode(_ object: [String: JSONValue]) throws -> ConnectMode? {
        guard let text = try optionalString(object, "connectMode") else { return nil }
        guard let mode = ConnectMode(rawValue: text) else {
            throw ProfilePortabilityError.invalidField(
                field: "connectMode",
                value: text,
                reason: "it must be either \"singleUser\" or \"multiUser\"."
            )
        }
        return mode
    }
}

// MARK: - Serialization

private extension PortableProfileDocument {
    /// The document as a JSON-serializable dictionary.
    ///
    /// Built by hand rather than through `Encodable` so optional-versus-absent is explicit: the
    /// schema distinguishes an absent `connectMode` (meaning "legacy profile, single user") from a
    /// present one, and a synthesized encoder would not preserve that distinction on re-export.
    public func jsonObject() -> [String: JSONValue] {
        var object: [String: JSONValue] = unknownFields
        object["schemaVersion"] = .number(Double(schemaVersion))
        object["id"] = .string(id.uuidString.lowercased())
        object["name"] = .string(name)
        object["ssoStartUrl"] = .string(ssoStartUrl)
        object["ssoRegion"] = .string(ssoRegion)
        object["accountId"] = .string(accountId)
        object["roleName"] = .string(roleName)
        object["resourceRegion"] = .string(resourceRegion)
        object["instanceTagKey"] = .string(instanceTagKey)
        object["instanceTagValue"] = .string(instanceTagValue)
        object["connectAction"] = .string(connectAction)
        object["localPort"] = .number(Double(localPort))
        object["remotePort"] = .number(Double(remotePort))

        // Absent stays absent. Writing an explicit null would be schema-valid for secretId but would
        // turn "this profile predates connectMode" into "this profile chose null", which is a
        // different statement and the exact thing AC-03 round-trips.
        if let connectMode { object["connectMode"] = .string(connectMode.rawValue) }
        if let agentRemotePort { object["agentRemotePort"] = .number(Double(agentRemotePort)) }
        if let secretId { object["secretId"] = .string(secretId) }

        return object
    }
}

/// A minimal JSON value, used only to carry unknown optional properties across a round-trip.
///
/// `Any` would have done the job in-process, but it is not `Sendable` or `Equatable`, and this
/// travels through a domain value that is both. It is also what keeps the import path off
/// `JSONSerialization`, whose `NSNumber` bridging (a `CFBoolean` check to tell `true` from `1`) is
/// Darwin-shaped — the wrong dependency for the type whose entire purpose is portability.
public enum JSONValue: Codable, Equatable, Sendable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null
    case array([JSONValue])
    case object([String: JSONValue])

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            self = .object(try container.decode([String: JSONValue].self))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .null:               try container.encodeNil()
        case let .bool(value):    try container.encode(value)
        case let .number(value):
            // Emit whole numbers as integers so ports and versions round-trip as `8443`, not `8443.0`.
            if value == value.rounded(), abs(value) < 1e15 {
                try container.encode(Int(value))
            } else {
                try container.encode(value)
            }
        case let .string(value):  try container.encode(value)
        case let .array(value):   try container.encode(value)
        case let .object(value):  try container.encode(value)
        }
    }
}

File: windows/src/SSMConnect.Domain/PortableProfile.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace SSMConnect.Domain;

/// <summary>
/// The portable connection-profile document (specification §6.1,
/// <c>contracts/connection-profile.schema.json</c>).
/// </summary>
/// <remarks>
/// Deliberately a separate type from <see cref="ConnectionProfile"/> rather than serialization
/// attributes on it. The native model is this client's private storage shape and is free to change;
/// this document is a cross-client contract and is not. They already disagree in ways that would
/// have shipped a schema-violating export: System.Text.Json writes enums as <b>numbers</b> by
/// default, so a naive export emits a numeric connectAction where the schema demands the string
/// <c>dcvViewer</c>, and the document carries a schemaVersion the native model has no reason to hold.
/// </remarks>
public sealed record PortableProfileDocument
{
    /// <summary>The only version this client writes, and the only one it accepts.</summary>
    public const int CurrentSchemaVersion = 1;

    /// <summary>The one connect action v1 defines.</summary>
    public const string DcvViewerAction = "dcvViewer";

    public int SchemaVersion { get; init; } = CurrentSchemaVersion;

    public required Guid Id { get; init; }
    public required string Name { get; init; }
    public required string SsoStartUrl { get; init; }
    public required string SsoRegion { get; init; }
    public required string AccountId { get; init; }
    public required string RoleName { get; init; }
    public required string ResourceRegion { get; init; }
    public required string InstanceTagKey { get; init; }
    public required string InstanceTagValue { get; init; }
    public required int LocalPort { get; init; }
    public required int RemotePort { get; init; }

    /// <summary>
    /// Null means the key was absent, and it must stay absent on re-export. The profile the
    /// shipping macOS app has stored predates this field, so writing a resolved singleUser would
    /// silently rewrite the user's document.
    /// </summary>
    public ConnectMode? ConnectMode { get; init; }

    /// <summary>Null means absent, for the same reason as <see cref="ConnectMode"/>.</summary>
    public int? AgentRemotePort { get; init; }

    public string? SecretId { get; init; }

    public string ConnectAction { get; init; } = DcvViewerAction;

    /// <summary>
    /// Optional properties this client does not model, preserved verbatim so a document written by
    /// a newer client survives a round-trip through this one.
    /// </summary>
    /// <remarks>
    /// Preservation is not unconditional. A document carrying any key on the schema's credential
    /// deny-list is rejected before it reaches here, so this can never become a smuggling channel.
    /// </remarks>
    public IReadOnlyDictionary<string, JsonNode?> UnknownFields { get; init; } =
        new Dictionary<string, JsonNode?>(StringComparer.Ordinal);

    public static PortableProfileDocument FromProfile(
        ConnectionProfile profile, IReadOnlyDictionary<string, JsonNode?>? unknownFields = null) => new()
        {
            Id = profile.Id,
            Name = profile.Name,
            SsoStartUrl = profile.SsoStartUrl,
            SsoRegion = profile.SsoRegion,
            AccountId = profile.AccountId,
            RoleName = profile.RoleName,
            ResourceRegion = profile.ResourceRegion,
            InstanceTagKey = profile.InstanceTagKey,
            InstanceTagValue = profile.InstanceTagValue,
            LocalPort = profile.LocalPort,
            RemotePort = profile.RemotePort,
            ConnectMode = profile.ConnectMode,
            AgentRemotePort = profile.AgentRemotePort,
            SecretId = profile.SecretId,
            UnknownFields = unknownFields ?? new Dictionary<string, JsonNode?>(StringComparer.Ordinal),
        };

    public ConnectionProfile ToProfile() => new()
    {
        Id = Id,
        Name = Name,
        SsoStartUrl = SsoStartUrl,
        SsoRegion = SsoRegion,
        AccountId = AccountId,
        RoleName = RoleName,
        ResourceRegion = ResourceRegion,
        InstanceTagKey = InstanceTagKey,
        InstanceTagValue = InstanceTagValue,
        LocalPort = LocalPort,
        RemotePort = RemotePort,
        ConnectMode = ConnectMode,
        AgentRemotePort = AgentRemotePort,
        SecretId = SecretId,
    };
}

/// <summary>
/// A portable document could not be read. Every case names the offending field and value, for the
/// same reason <see cref="ProfileConfigurationException"/> does: an import failure the user cannot
/// act on is barely better than a silent one.
/// </summary>
public sealed class ProfilePortabilityException(string message) : ConnectionException(message)
{
    public override ErrorCategory Category => ErrorCategory.Configuration;

    public static ProfilePortabilityException MalformedJson() =>
        new("The file is not valid JSON.");

    public static ProfilePortabilityException NotAnObject() =>
        new("The file does not contain a profile object.");

    public static ProfilePortabilityException UnsupportedSchemaVersion(int found) =>
        new($"This profile uses schema version {found}, which this version of SSM Connect does not " +
            "understand. Update SSM Connect and try again.");

    public static ProfilePortabilityException ForbiddenField(string key) =>
        new($"The profile contains \"{key}\", which may carry credentials. A profile document must " +
            "never contain authentication material, so it was rejected rather than imported.");

    public static ProfilePortabilityException MissingField(string key) =>
        new($"The profile is missing the required field \"{key}\".");

    public static ProfilePortabilityException InvalidField(string field, string value, string reason) =>
        new($"The profile field \"{field}\" has the value \"{value}\", but {reason}");
}

/// <summary>Reads and writes the portable profile document.</summary>
public static class ProfilePortability
{
    /// <summary>
    /// Keys the schema's security clause forbids outright: live authentication material and
    /// machine-local paths. Rejected even though they would otherwise be unknown optional
    /// properties, because preserving them is precisely the failure mode to avoid.
    /// </summary>
    public static IReadOnlySet<string> ForbiddenKeys { get; } = new HashSet<string>(StringComparer.Ordinal)
    {
        "accessKeyId", "secretAccessKey", "sessionToken", "credentials",
        "accessToken", "refreshToken", "ssoAccessToken",
        "password", "dcvPassword", "secretValue", "authToken", "presignedUrl",
        "sessionResponse", "ssmSession", "connectionFilePath", "tempFilePath",
    };

    private static readonly HashSet<string> KnownKeys = new(StringComparer.Ordinal)
    {
        "schemaVersion", "id", "name", "ssoStartUrl", "ssoRegion", "accountId", "roleName",
        "resourceRegion", "instanceTagKey", "instanceTagValue", "connectMode", "connectAction",
        "localPort", "remotePort", "agentRemotePort", "secretId",
    };

    private static readonly JsonSerializerOptions WriteOptions = new() { WriteIndented = true };

    /// <summary>
    /// Encodes a profile as a portable document. Keys are written in a stable order and the output
    /// is indented, because an exported profile is a file people diff and review.
    /// </summary>
    public static string Export(
        ConnectionProfile profile, IReadOnlyDictionary<string, JsonNode?>? unknownFields = null) =>
        Export(PortableProfileDocument.FromProfile(profile, unknownFields));

    public static string Export(PortableProfileDocument document)
    {
        var root = new JsonObject
        {
            ["schemaVersion"] = document.SchemaVersion,
            ["id"] = document.Id.ToString("D"),
            ["name"] = document.Name,
            ["ssoStartUrl"] = document.SsoStartUrl,
            ["ssoRegion"] = document.SsoRegion,
            ["accountId"] = document.AccountId,
            ["roleName"] = document.RoleName,
            ["resourceRegion"] = document.ResourceRegion,
            ["instanceTagKey"] = document.InstanceTagKey,
            ["instanceTagValue"] = document.InstanceTagValue,
            ["connectAction"] = document.ConnectAction,
            ["localPort"] = document.LocalPort,
            ["remotePort"] = document.RemotePort,
        };

        // Absent stays absent. Resolving a default while writing would rewrite the user's document
        // and would make a round-trip test pass against synthetic data only.
        if (document.ConnectMode is ConnectMode mode)
        {
            root["connectMode"] = mode == Domain.ConnectMode.MultiUser ? "multiUser" : "singleUser";
        }

        if (document.AgentRemotePort is int agentPort)
        {
            root["agentRemotePort"] = agentPort;
        }

        if (document.SecretId is not null)
        {
            root["secretId"] = document.SecretId;
        }

        foreach ((string key, JsonNode? value) in document.UnknownFields)
        {
            root[key] = value?.DeepClone();
        }

        var ordered = new JsonObject();
        foreach (string key in root.Select(pair => pair.Key).OrderBy(key => key, StringComparer.Ordinal).ToList())
        {
            ordered[key] = root[key]?.DeepClone();
        }

        return ordered.ToJsonString(WriteOptions);
    }

    /// <summary>Reads a portable document, validating it the way the schema does.</summary>
    public static PortableProfileDocument Import(string json)
    {
        JsonNode? node;
        try
        {
            node = JsonNode.Parse(json);
        }
        catch (JsonException)
        {
            throw ProfilePortabilityException.MalformedJson();
        }

        if (node is not JsonObject root)
        {
            throw ProfilePortabilityException.NotAnObject();
        }

        // Version first: a newer document must fail with an actionable message rather than be
        // partially understood.
        int version = RequiredInt(root, "schemaVersion");
        if (version != PortableProfileDocument.CurrentSchemaVersion)
        {
            throw ProfilePortabilityException.UnsupportedSchemaVersion(version);
        }

        // Reject before reading anything else, so a document carrying a secret is never partially
        // materialised.
        foreach (string key in root.Select(pair => pair.Key).ToList())
        {
            if (ForbiddenKeys.Contains(key))
            {
                throw ProfilePortabilityException.ForbiddenField(key);
            }
        }

        string idText = RequiredString(root, "id");
        if (!Guid.TryParse(idText, out Guid id))
        {
            throw ProfilePortabilityException.InvalidField("id", idText, "it is not a UUID.");
        }

        ConnectMode? mode = null;
        if (root["connectMode"] is JsonNode modeNode && modeNode.GetValueKind() != JsonValueKind.Null)
        {
            string modeText = modeNode.GetValue<string>();
            mode = modeText switch
            {
                "singleUser" => Domain.ConnectMode.SingleUser,
                "multiUser" => Domain.ConnectMode.MultiUser,
                _ => throw ProfilePortabilityException.InvalidField(
                    "connectMode", modeText, "it must be singleUser or multiUser."),
            };
        }

        string action = PortableProfileDocument.DcvViewerAction;
        if (root["connectAction"] is JsonNode actionNode && actionNode.GetValueKind() != JsonValueKind.Null)
        {
            action = actionNode.GetValue<string>();
            if (action != PortableProfileDocument.DcvViewerAction)
            {
                throw ProfilePortabilityException.InvalidField(
                    "connectAction", action,
                    $"this version only supports {PortableProfileDocument.DcvViewerAction}.");
            }
        }

        string? secretId = null;
        if (root["secretId"] is JsonNode secretNode && secretNode.GetValueKind() != JsonValueKind.Null)
        {
            secretId = secretNode.GetValue<string>();
        }

        if (mode == Domain.ConnectMode.MultiUser && secretId is not null)
        {
            throw ProfilePortabilityException.InvalidField(
                "secretId", secretId,
                "a multi-user workstation authenticates by identity and has no shared password.");
        }

        var unknown = new Dictionary<string, JsonNode?>(StringComparer.Ordinal);
        foreach ((string key, JsonNode? value) in root.ToList())
        {
            if (!KnownKeys.Contains(key))
            {
                unknown[key] = value?.DeepClone();
            }
        }

        return new PortableProfileDocument
        {
            SchemaVersion = version,
            Id = id,
            Name = RequiredNonEmptyString(root, "name"),
            SsoStartUrl = RequiredNonEmptyString(root, "ssoStartUrl"),
            SsoRegion = RequiredRegion(root, "ssoRegion"),
            AccountId = RequiredAccountId(root),
            RoleName = RequiredNonEmptyString(root, "roleName"),
            ResourceRegion = RequiredRegion(root, "resourceRegion"),
            InstanceTagKey = RequiredNonEmptyString(root, "instanceTagKey"),
            InstanceTagValue = RequiredString(root, "instanceTagValue"),
            LocalPort = RequiredPort(root, "localPort"),
            RemotePort = RequiredPort(root, "remotePort"),
            ConnectMode = mode,
            AgentRemotePort = OptionalPort(root, "agentRemotePort"),
            SecretId = secretId,
            ConnectAction = action,
            UnknownFields = unknown,
        };
    }

    public static ConnectionProfile ImportProfile(string json) => Import(json).ToProfile();

    private static JsonNode RequiredValue(JsonObject root, string key) =>
        root[key] ?? throw ProfilePortabilityException.MissingField(key);

    private static string RequiredString(JsonObject root, string key)
    {
        JsonNode value = RequiredValue(root, key);
        if (value.GetValueKind() != JsonValueKind.String)
        {
            throw ProfilePortabilityException.InvalidField(key, value.ToJsonString(), "it must be a string.");
        }

        return value.GetValue<string>();
    }

    private static string RequiredNonEmptyString(JsonObject root, string key)
    {
        string value = RequiredString(root, key);
        if (value.Length == 0)
        {
            throw ProfilePortabilityException.InvalidField(key, value, "it must not be empty.");
        }

        return value;
    }

    private static int RequiredInt(JsonObject root, string key)
    {
        JsonNode value = RequiredValue(root, key);
        if (value.GetValueKind() != JsonValueKind.Number || !value.AsValue().TryGetValue(out int parsed))
        {
            throw ProfilePortabilityException.InvalidField(key, value.ToJsonString(), "it must be a whole number.");
        }

        return parsed;
    }

    private static string RequiredRegion(JsonObject root, string key)
    {
        string value = RequiredString(root, key);
        if (!AwsRegion.IsValid(value))
        {
            throw ProfilePortabilityException.InvalidField(key, value, "it is not a valid AWS region.");
        }

        return value;
    }

    private static string RequiredAccountId(JsonObject root)
    {
        string value = RequiredString(root, "accountId");
        if (value.Length != 12 || !value.All(char.IsAsciiDigit))
        {
            throw ProfilePortabilityException.InvalidField(
                "accountId", value, "an AWS account ID is exactly 12 digits.");
        }

        return value;
    }

    private static int RequiredPort(JsonObject root, string key)
    {
        int value = RequiredInt(root, key);
        if (value is < 1 or > 65535)
        {
            throw ProfilePortabilityException.InvalidField(
                key, value.ToString(CultureInfo.InvariantCulture), "a port must be between 1 and 65535.");
        }

        return value;
    }

    private static int? OptionalPort(JsonObject root, string key)
    {
        if (root[key] is not JsonNode node || node.GetValueKind() == JsonValueKind.Null)
        {
            return null;
        }

        return RequiredPort(root, key);
    }
}
